src/ray/core_worker/core_worker.cc

```diff
@@ -3016,40 +3016,54 @@ Status CoreWorker::ReportGeneratorItemReturns(
 
-  waiter->IncrementObjectGenerated();
+  if (waiter) {
+    waiter->IncrementObjectGenerated();
+  }
+
   client->ReportGeneratorItemReturns(
       request,
       [waiter, generator_id, item_index](
           const Status &status, const rpc::ReportGeneratorItemReturnsReply &reply) {
         RAY_LOG(DEBUG) << "ReportGeneratorItemReturns replied. " << generator_id
                        << "index: " << item_index << ". total_consumed_reported: "
                        << reply.total_num_object_consumed();
-        RAY_CHECK(waiter != nullptr);
-        RAY_LOG(DEBUG) << "Total object consumed: " << waiter->TotalObjectConsumed()
-                       << ". Total object generated: " << waiter->TotalObjectGenerated();
-        if (status.ok()) {
-          /// Since unary gRPC requests are not ordered, it is possible the stale
-          /// total value can be replied. Since total object consumed only can
-          /// increment, we always choose the larger value here.
-          waiter->UpdateTotalObjectConsumed(
-              std::max(waiter->TotalObjectConsumed(), reply.total_num_object_consumed()));
-        } else {
-          // TODO(sang): Handle network error more gracefully.
-          // If the request fails, we should just resume until task finishes without
-          // backpressure.
-          waiter->UpdateTotalObjectConsumed(waiter->TotalObjectGenerated());
-          RAY_LOG(WARNING) << "Failed to send the object ref.";
+        if (waiter) {
+          RAY_LOG(DEBUG) << "Total object consumed: " << waiter->TotalObjectConsumed()
+                         << ". Total object generated: "
+                         << waiter->TotalObjectGenerated();
+          if (status.ok()) {
+            /// Since unary gRPC requests are not ordered, it is possible the stale
+            /// total value can be replied. Since total object consumed only can
+            /// increment, we always choose the larger value here.
+            waiter->UpdateTotalObjectConsumed(std::max(
+                waiter->TotalObjectConsumed(), reply.total_num_object_consumed()));
+          } else {
+            // TODO(sang): Handle network error more gracefully.
+            // If the request fails, we should just resume until task finishes without
+            // backpressure.
+            waiter->UpdateTotalObjectConsumed(waiter->TotalObjectGenerated());
+            RAY_LOG(WARNING) << "Failed to send the object ref.";
+          }
         }
       });
+
   // Backpressure if needed. See task_manager.h and search "backpressure" for protocol
   // details.
-  return waiter->WaitUntilObjectConsumed(/*check_signals*/ [this]() {
+  if (waiter) {
+    return waiter->WaitUntilObjectConsumed(/*check_signals*/ [this]() {
+      if (options_.check_signals) {
+        return options_.check_signals();
+      } else {
+        return Status::OK();
+      }
+    });
+  } else {
     if (options_.check_signals) {
```

Comment:
I think for else case, we can just return Status::OK. (if you look at the impl of WaitUntilObjectConsumed, that's the behavior now). check_signals is to avoid ignoring python signals while cpp is using the thread (which doesn't happen in this case)

Reply:
I have revisited it to avoid duplication, essentially sharing the callbkac

Reply:
actually we don't need to check signal if there's no waiter. The purpose of checking signal is when the thread is blocked inside cpp (e.g., due to locks or sleep), it cannot check python interrupt. But if we don't have the waiter there's no such problem